### PR TITLE
hook-digest: Add hook for computing digests

### PR DESCRIPTION
**Pull request by PJungkamp**
Resolves: #702

## ToDo

- [x] Give reasonable error messages instead of asserts.
- [x] ~~Should I just remove the `villas-ts-sign.sh` and `villas-ts-verify.py` tools commit?~~

### CODE REVIEW

**Comment by PJungkamp** [comment]
@stv0g I'm not quite sure how to configure the `flags` of a `Format`. E.g for the `villas.human` format.

I see that the constructor of a `Format` is passed a set of flags. See https://github.com/VILLASframework/node/blob/b9bad1219d69b0e601342ba5c81e4c712a71d286/lib/format.cpp#L60-L64
But I'm instantiating the Format using `FormatFactory::make` which does not take flags for an argument.

**Comment by PJungkamp** [comment]
It's kind of unfortunate that I had to transition to older APIs to support `OpenSSL 1.1.1`. Especially since `OpenSSL 1.1.1` reaches its end of life in less than two months.

File: doc/openapi/components/schemas/config/hooks/digest.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
+# yaml-language-server: $schema=http://json-schema.org/draft-07/schema
+---
+
+allOf:
+- type: object
+  required:
+    - "uri"
+    - "mode"
+    - "interval"
+    - "algorithm"
```
Comment (stv0g) [inline doc/openapi/components/schemas/config/hooks/digest.yaml:10]:
```suggestion
  required:
  - uri
  - mode
  - interval
  - algorithm
```

File: lib/hooks/digest.cpp [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+#include <villas/timing.hpp>
+#include <villas/exceptions.hpp>
+#include <openssl/evp.h>
+#include <openssl/crypto.h>
+#include <optional>
+#include <variant>
+#include <string>
+#include <iterator>
+#include <memory>
+#include <fmt/format.h>
```
Comment (stv0g) [inline lib/hooks/digest.cpp:21]: Include order

```suggestion
#include <optional>
#include <variant>
#include <string>
#include <iterator>
#include <memory>

#include <openssl/evp.h>
#include <openssl/crypto.h>
#include <fmt/format.h>

#include <villas/utils.hpp>
#include <villas/format.hpp>
#include <villas/hook.hpp>
#include <villas/sample.hpp>
#include <villas/timing.hpp>
#include <villas/exceptions.hpp>
```

File: lib/hooks/digest.cpp [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+
+namespace villas {
+namespace node {
+
+enum Mode {
+	Timestamp,
+	Sequence,
+};
+
+static void FILE_free(FILE *f) {
```
Comment (stv0g) [inline lib/hooks/digest.cpp:31]:
```suggestion
static void FILE_free(FILE *f)
{
```

File: lib/hooks/digest.cpp [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+	}
+};
+
+/* Register hook */
+static char n[] = "digest";
+static char d[] = "Calculate the digest for a range of samples";
+static HookPlugin<DigestHook, n, d, (int) Hook::Flags::NODE_READ | (int) Hook::Flags::NODE_WRITE | (int) Hook::Flags::PATH, 20> p;
+
+} /* namespace node */
+} /* namespace villas */
```
Comment (stv0g) [inline lib/hooks/digest.cpp:224]: I am trying to refactor the codebase to use C++ style comments (`//`) rather C-style (`/* */`)

```suggestion
} // namespace node
} // namespace villas
```

File: lib/hooks/digest.cpp [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+	void stop() {
+		Hook::stop();
+
+		current_interval = std::nullopt;
+		if (!EVP_MD_CTX_reset(md_ctx.get()))
+			throw RuntimeError { "Could not reset digest context" };
+	}
+};
+
+/* Register hook */
```
Comment (stv0g) [inline lib/hooks/digest.cpp:218]:
```suggestion
// Register hook
```

File: lib/hooks/digest.cpp [hunk 6]
```diff
@@ -0,0 +1,224 @@
+/** Digest hook.
+ *
+ * @author Philipp Jungkamp <philipp.jungkamp@opal-rt.com>
+ * @copyright 2023, OPAL-RT Germany GmbH
+ * @license Apache 2.0
+ *********************************************************************************/
```
Comment (stv0g) [inline lib/hooks/digest.cpp:6]: Part of a cleanup which I am currently performing in another PR
```suggestion
/* Digest hook.
 *
 * Author: Philipp Jungkamp <philipp.jungkamp@opal-rt.com>
 * SPDX-FileCopyrightText: 2023 OPAL-RT Germany GmbH
 * SPDX-License-Identifier: Apache-2.0
 */
```

File: doc/openapi/components/schemas/config/hooks/_digest.yaml [hunk 7]
```diff
@@ -0,0 +1,3 @@
+allOf:
```
Comment (stv0g) [inline doc/openapi/components/schemas/config/hooks/_digest.yaml:1]:
```suggestion
# yaml-language-server: $schema=http://json-schema.org/draft-07/schema
# SPDX-FileCopyrightText: 2023 OPAL-RT Germany GmbH
# SPDX-License-Identifier: Apache-2.0
---
allOf:
```

File: etc/examples/hooks/digest.conf [hunk 8]
```diff
@@ -0,0 +1,33 @@
+@include "hook-nodes.conf"
```
Comment (stv0g) [inline etc/examples/hooks/digest.conf:1]:
```suggestion
# SPDX-FileCopyrightText: 2023 OPAL-RT Germany GmbH
# SPDX-License-Identifier: Apache-2.0

@include "hook-nodes.conf"
```

File: lib/hooks/digest.cpp [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+	Mode mode;
+	std::variant<int, double> interval;
+
+	FILE *file;
+	std::optional<int> current_interval;
+	std::string md_string_buffer;
+	Format *format;
+	std::string format_buffer;
+
+	bool updateInterval(Sample *smp) {
```
[the remark this answers is not in the document]
Comment (stv0g) [reply lib/hooks/digest.cpp:55]: Should be pass `const Sample *smp`?

File: lib/hooks/digest.cpp [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+		auto inserter = std::back_inserter(md_string_buffer);
+
+		fmt::format_to(inserter, "{}:", digestname);
+		for (unsigned int i = 0; i < md_size; ++i)
+			fmt::format_to(inserter, "{:02X}", md_value[i]);
+
+		fprintf(file, "%d %s\n", current_interval.value(), md_string_buffer.c_str());
+	}
+
+	void updateDigest(Sample *smp) {
```
[the remark this answers is not in the document]
Comment (stv0g) [reply lib/hooks/digest.cpp:99]: Should be pass `const Sample *smp`?

**Review by stv0g — CHANGES_REQUESTED** [review]
Thanks :) Looks good 👍🏻

**Comment by stv0g** [comment]
> Should I just use the black formatter?

**Comment by PJungkamp** [comment]
> But after our discussion yesterday, we should also consider moving the timestamping code to the SEGuRo Platform repo.

I'll probably just remove the script in the `tools` directory.

File: lib/hooks/frame.cpp [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+        auto next_millisecond_of_second =
+            next_smp->ts.origin.tv_nsec / 1'000'000;
+        auto last_interval_of_second =
+            (last_millisecond_of_second + interval - offset) / interval;
+        auto next_interval_of_second =
+            (next_millisecond_of_second + interval - offset) / interval;
+        changed = last_interval_of_second != next_interval_of_second;
+        break;
+      }
+      }
```
[the remark this answers is not in the document]
Comment (PJungkamp) [reply lib/hooks/frame.cpp:163]: LLVM codestyle for switch case with scoped cases is weird.

File: doc/openapi/components/schemas/config/hooks/frame.yaml [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+      description: The trigger for new frames.
+      type: string
+      default: sequence
+      enum:
+      - sequence
+      - timestamp
+    unit:
+      description: The type of a timestamp trigger.
+      type: string
+      enum:
```
[the remark this answers is not in the document]
Comment (PJungkamp) [reply doc/openapi/components/schemas/config/hooks/frame.yaml:18]: unit has no default.

File: doc/openapi/components/schemas/config/hooks/digest.yaml [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+    uri:
+      description: The output file for digests.
+      example: digest.txt
+      type: string
+      format: uri
+
+    algorithm:
+      description: The algorithm used for calculating digests.
+      example: sha256
+      type: string
```
[the remark this answers is not in the document]
Comment (PJungkamp) [reply doc/openapi/components/schemas/config/hooks/digest.yaml:23]: No, I wanted this to be explicit.

**Comment by stv0g** [comment]
- [ ] Default values OpenAPI schema of framing hook
- [ ] `json_flags` vs `flags_json`
- [ ] Remove initialization `last_sample->flags`
- [ ] Integration tests for new hooks: reorder, frame, digest
- [ ] path_source proper fixups for `muxed_sample->flags`

**Comment by PJungkamp** [comment]
@stv0g Could you have a look at the failing `hook-digest` integration test. I can't reproduce the error locally.

**Review by stv0g — APPROVED** [review]

**Pull request merged**